File: isma-intg-api/src/main/java/ru/nstu/isma/intg/api/methods/AccuracyIntgController.java
package ru.nstu.isma.intg.api.methods;

import ru.nstu.isma.intg.api.solvers.DaeSystemStepSolver;

import java.io.Serializable;

/**
 * @author Maria Nasyrova
 * @since 25.02.2015
 */
public abstract class AccuracyIntgController extends IntgController {

    private double accuracy;

    public double getAccuracy() {
        return accuracy;
    }

    public void setAccuracy(double accuracy) {
        this.accuracy = accuracy;
    }

    public abstract AccuracyResults tune(
            IntgPoint fromPoint, double[][] stages, DaeSystemStepSolver stepSolver);

    protected abstract double getLocalErrorEstimation(double[] deStages);

    protected abstract double getActualAccuracy(double step, double localError);

    protected abstract double getQ(double accuracy, double actualAccuracy);


    public static class AccuracyResults implements Serializable {
        private final double tunedStep;
        private final double[][] tunedStages;

        public AccuracyResults(double tunedStep, double[][] tunedStages) {
            this.tunedStep = tunedStep;
            this.tunedStages = tunedStages;
        }

        public double getTunedStep() {
            return tunedStep;
        }

        public double[][] getTunedStages() {
            return tunedStages;
        }
    }
}

File: isma-intg-api/src/main/kotlin/ru/nstu/isma/intg/api/methods/AccuracyIntgController.kt
package ru.nstu.isma.intg.api.methods

import ru.nstu.isma.intg.api.solvers.DaeSystemStepSolver
import java.io.Serializable

abstract class AccuracyIntgController: IntgController() {
    var accuracy = 0.0

    abstract fun tune(
        fromPoint: IntgPoint, stages: Array<DoubleArray>, stepSolver: DaeSystemStepSolver
    ): AccuracyResults

    protected abstract fun getLocalErrorEstimation(deStages: DoubleArray): Double

    protected abstract fun getActualAccuracy(step: Double, localError: Double): Double

    protected abstract fun getQ(accuracy: Double, actualAccuracy: Double): Double


    class AccuracyResults(val tunedStep: Double, val tunedStages: Array<DoubleArray>) : Serializable
}